File: crawler-scala/src/main/scala/com/newspulse/crawler/model/Article.scala
package com.newspulse.crawler.model

import play.api.libs.json._
import java.time.Instant

/**
 * Represents a news article
 */
case class Article(
  id: String,
  url: String,
  title: String,
  description: Option[String],
  content: String,
  author: Option[String],
  source: String,
  category: Option[String],
  tags: List[String],
  imageUrl: Option[String],
  publishTime: Option[Instant],
  crawlTime: Instant
)

object Article:
  given Format[Instant] = new Format[Instant]:
    def reads(json: JsValue): JsResult[Instant] = 
      json.validate[Long].map(Instant.ofEpochMilli)
    def writes(instant: Instant): JsValue = 
      JsNumber(instant.toEpochMilli)

  given Format[Article] = Json.format[Article]
  
  def generateId(url: String): String =
    val digest = java.security.MessageDigest.getInstance("MD5")
    val hash = digest.digest(url.getBytes("UTF-8"))
    hash.map("%02x".format(_)).mkString

File: api-springboot/src/main/java/com/newspulse/api/model/Article.java
package com.newspulse.api.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.time.Instant;
import java.util.List;

/**
 * Article model for API responses
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Article implements Serializable {

    private static final long serialVersionUID = 1L;

    private String id;
    private String url;
    private String title;
    private String description;
    private String content;
    private String author;
    private String source;
    private String category;
    private List<String> tags;
    private String imageUrl;
    private Instant publishTime;
    private Instant crawlTime;
    private String language;
    private Integer wordCount;
    private Integer clusterId;
    private Double trendingScore;

    // For search results
    private Double score;
    private String highlightTitle;
    private String highlightContent;
}
